### PR TITLE
Fix Union Types Import Issue

### PR DESCRIPTION
**Pull request by FrankEssenberger**
This is an attempt to fix the bug reported here: #5467 
The problem was that the union types in typescrit e.g. `TypeA | TypeB` represented by the `oneOf` keyword in the service definition lead to not valid code:
```
import{TypeA | TypeB} from '../models'
```
you can either use a comma in the import or split into separate imports. I did the latter one because I thought this could be more solid for models in different files and paths potentially. If they are not all exported from the same path.

I adjusted it for three typescript generators:
- axios (this I where I was effected)
- node
- angular

Then I stopped and thought it is a good point to start the discussion before I adjust it for all the other Typescript generators. 

Since I am new I would highly appreciate some help regarding the following issues:
- [ ] How to run the code formatter. When I run `mvn formatter:format` I get hundreds of code changes.
- [ ] How to run the tests. Is `mvn test` fine? 
- [ ] Not sure about the locations where I put the tests and so on. Perhaps this kind of test has to go somewhere else.

### PR checklist

- [x] Read the [contribution guidelines](https://github.com/openapitools/openapi-generator/blob/master/CONTRIBUTING.md).
~~If contributing template-only or documentation-only changes which will change sample output, [build the project](https://github.com/OpenAPITools/openapi-generator#14---build-projects) beforehand.~~
- [ ] Run the shell script `./bin/generate-samples.sh`to update all Petstore samples related to your fix. This is important, as CI jobs will verify _all_ generator outputs of your HEAD commit as it would merge with master. These must match the expectations made by your contribution. You may regenerate an individual generator by passing the relevant config(s) as an argument to the script, for example `./bin/generate-samples.sh bin/config/java*`. For Windows users, please run the script in [Git BASH](https://gitforwindows.org/).
- [ ] File the PR against the [correct branch](https://github.com/OpenAPITools/openapi-generator/wiki/Git-Branches): `master`
- [x] Copy the [technical committee](https://github.com/openapitools/openapi-generator/#62---openapi-generator-technical-committee) to review the pull request if your PR is targeting a particular programming language.

### CODE REVIEW

**Comment by FrankEssenberger** [comment]
@TiFu (2017/07) @taxpon (2017/07) @sebastianhaas (2017/07) @kenisteward (2017/07) @Vrolijkx (2017/09) @macjohnny (2018/01) @topce (2018/10) @akehir (2019/07) @petejohansonxo (2019/11) @amakhrov (2020/02)

File: modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/AbstractTypeScriptClientCodegen.java
Comment (macjohnny) [inline modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/AbstractTypeScriptClientCodegen.java:228]: please add some docs to describe what this does, with an example

File: modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/AbstractTypeScriptClientCodegen.java
Comment (macjohnny) [inline modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/AbstractTypeScriptClientCodegen.java:239]: docs

File: modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/AbstractTypeScriptClientCodegen.java
Comment (macjohnny) [inline modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/AbstractTypeScriptClientCodegen.java:260]: @amakhrov @wing328  do you think such a complicated logic is necessary, or could this be solved elsewhere?

Comment (amakhrov) [reply modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/AbstractTypeScriptClientCodegen.java:260]: It looks quite similar to the existing logic in TypeScriptAngularClientCodegen (see `postProcessAllModels` and `parseImports`)

Maybe we should lift the logic from AngularClient to AbstractTypescriptClient instead of pretty much repeating it?

The root problem, as I see it, is that `toModelImport` takes a string as an argument rather than a model and returns a string instead of a list. It makes it very limited in terms of handling composite models. But changing that would require a change in Generator and CodegenConfig interfaces. And generally means more work, of course.

**Comment by macjohnny** [comment]
@FrankEssenberger thanks for your contribution!

File: modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/AbstractTypeScriptClientCodegen.java
Comment (FrankEssenberger) [reply modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/AbstractTypeScriptClientCodegen.java:228]: I added the doc to the new method `toModelImportMap`

File: modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/AbstractTypeScriptClientCodegen.java
Comment (FrankEssenberger) [reply modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/AbstractTypeScriptClientCodegen.java]: removed the method.

**Comment by FrankEssenberger** [comment]
I adjusted my approach and created a second method `Map<String,String> toModelImportMap(String input)` which will give the map between the fully quantified model for import and the initial name. For most cases this map will have only a single pair. However, for the typescript case if a type like "classA | classB" is found it will have two entries. 

This seemed much cleaner then to fiddle around with the imports in a post processing step.

**Comment by FrankEssenberger** [comment]
I would like to run the formatter so that the diff looks better. Please let me know how to execute this and I am also happy for further feedback.

**Comment by wing328** [comment]
cc @OpenAPITools/generator-core-team as the change impacts default codegen, generator

File: modules/openapi-generator/src/main/java/org/openapitools/codegen/DefaultGenerator.java
Comment (TiFu) [inline modules/openapi-generator/src/main/java/org/openapitools/codegen/DefaultGenerator.java:20]: Is `google.common.colllect` preferred over `java.util.*` in OAI generator?

Comment (FrankEssenberger) [reply modules/openapi-generator/src/main/java/org/openapitools/codegen/DefaultGenerator.java]: Just an old habit from other projects. I used the basic java methods to create new Lists and Maps.

**Comment by macjohnny** [comment]
@FrankEssenberger could you please merge the current master and re-generate the samples?

**Comment by FrankEssenberger** [comment]
I have done a `./bin/generate-samples.sh` and merged master. However, there were a lot of code changes in the samples which I commited. Not sure if this is correct?

**Comment by macjohnny** [comment]
@FrankEssenberger did you
- merge master
- `mvn clean package`
- re-generate the samples
?

**Comment by FrankEssenberger** [comment]
yes this is what I did. If I do this on the not forked repo I do not get the changes in the samples. Also the changes are reordering of dependencies for example:

```
#include "ApiResponse.h"	
#include "HttpContent.h"	
#include "Pet.h"	#include "Pet.h"
#include <cpprest/details/basic_types.h>
```
changes to
```
#include "Pet.h"	#include "Pet.h"
#include <cpprest/details/basic_types.h>
#include "ApiResponse.h"	
#include "HttpContent.h"	
```

For whatever reason. Is there a formatter task present?

**Comment by macjohnny** [comment]
Could it be your IDE that performs code cleanup when committing?

**Comment by FrankEssenberger** [comment]
ah, I just realized that with me map introduction I changed the order of imports. Most likely the hashmap has not a alphabetic ordering when I loop over the keys. I will fix this that the imports are order alphabetically.

**Comment by FrankEssenberger** [comment]
ok, now only the typescript related samples have changed which is plausible.

**Comment by FrankEssenberger** [comment]
looked at the failing tests and the missing imports are not correct. So I have broken something with my changes so that the for example the `Tag` and `Category` are not there anymore. But I have a local test running so I can fix this issue quickly I hope.

**Comment by FrankEssenberger** [comment]
so I reverted a change to the angular typescript generator. It seems that the imports come from two places for api and model files. I did not expect that.

**Comment by FrankEssenberger** [comment]
hm, I do not find good information on the failed test. All I see in the step is:
```
2020-07-12 11:31:33.713:INFO:oejs.AbstractNCSARequestLog:main: Opened /var/log/2020_07_12-requests.log
2020-07-12 11:31:33.743:INFO:oejs.ServerConnector:main: Started ServerConnector@1e8b7643{HTTP/1.1}{0.0.0.0:8080}
2020-07-12 11:31:33.745:INFO:oejs.Server:main: Started @5914ms

Exited with code exit status 1
CircleCI received exit code 1
```
which is rather unspecific. Are these tests known to be flaky? Should I trigger a rerun?

**Comment by macjohnny** [comment]
@FrankEssenberger I re-triggered the CI build, let's see

**Comment by macjohnny** [comment]
@FrankEssenberger can you merge the most recent master, again, please?

**Comment by FrankEssenberger** [comment]
Sure. Actually I planned on the weekend to do something. I wanted to comment out the tests I added to see if they cause problems.

**Comment by FrankEssenberger** [comment]
So it turned out that test should have the word `Test` somewhere in their name, I think this is some surefire default and that it is not a good idea to create files in a test and not clean them up. Since the tests are green now, it would be great to get finally the review done.

File: modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/AbstractTypeScriptClientCodegen.java
Comment (amakhrov) [inline modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/AbstractTypeScriptClientCodegen.java:256]: These two methods (toModelImportMap and toModelImport) look very similar (split by `|`, remove spaces, process each part). Should one method just delegate to the other? E.g. `toModelImport` could call `toModelImportMap` and then collect the resulting map back into a string.

Comment (FrankEssenberger) [reply modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/AbstractTypeScriptClientCodegen.java:250]: sure good point. I introduced small helper methods with a more descriptive name and used the map method inside the single entry one.

**Comment by FrankEssenberger** [comment]
The pipeline seems a bit flaky. I see in the ci build the error: 
```
[error] sbt.librarymanagement.ResolveException: Error downloading com.softwaremill.sttp.client:json4s_2.13:2.2.0
[error]   Not found
[error]   Not found
[error]   not found: /home/circleci/.ivy2/local/com.softwaremill.sttp.client/json4s_2.13/2.2.0/ivys/ivy.xml
[error]   download error: Caught java.net.ConnectException: Connection timed out (Connection timed out) (Connection timed out (Connection timed out)) while downloading https://repo1.maven.org/maven2/com/softwaremill/sttp/client/json4s_2.13/2.2.0/json4s_2.13-2.2.0.pom
[error] 	at lmcoursier.CoursierDependencyResolution.unresolvedWarningOrThrow(CoursierDependencyResolution.scala:249)
```

**Comment by FrankEssenberger** [comment]
And for the travis-ci I see the errorr:
```
 1) Isomorphic Fetch
       GET-Request:
     Error: Timeout of 10000ms exceeded. For async tests and hooks, ensure "done()" is called; if returning a Promise, ensure it resolves. (/home/travis/build/OpenAPITools/openapi-generator/samples/openapi3/client/petstore/typescript/tests/default/test/http/isomorphic-fetch.test.ts)
```
However, if I run the test locally it runs through so perhaps the timeout is reached for some strange reason in the pipeline.

**Comment by macjohnny** [comment]
@FrankEssenberger I re-started the CI builds.

**Comment by FrankEssenberger** [comment]
Cool, your restart helped. You have the magic green finger :-).

**Comment by macjohnny** [comment]
cc @OpenAPITools/generator-core-team for final approval, as the change impacts default codegen, generator

**Comment by FrankEssenberger** [comment]
merged master. Do you need something from my side?

**Comment by macjohnny** [comment]
@FrankEssenberger we need to wait for a final review a a member of the @OpenAPITools/generator-core-team

File: modules/openapi-generator/src/main/java/org/openapitools/codegen/DefaultGenerator.java
Comment (wing328) [inline modules/openapi-generator/src/main/java/org/openapitools/codegen/DefaultGenerator.java:1102]: Can you please add docstrings to explain what these new (private) functions do (e.g. avoid duplicates)?

We want to document all methods in the default codegen/generator class to make code easier to understand for new contributors. (I do not expect you to document all methods missing the docstring. The community can help with separate PRs)

Comment (FrankEssenberger) [reply modules/openapi-generator/src/main/java/org/openapitools/codegen/DefaultGenerator.java]: I added some doc and used a Set to better represent that the entries are unique for the other method.

**Comment by macjohnny** [comment]
@wing328 can we merge this?

**Comment by FelschR** [comment]
`typescript-fetch` has the same issue. Would be nice to fix it there as well.

**Comment by FrankEssenberger** [comment]
Yes, I put the change in the `AbstractTypeScriptClientCodegen.java` which should be the parent for all typescript generators I guess. In the test I piked three axios, node and angular and for all the union types were correctly handled.

**Comment by macjohnny** [comment]
@FrankEssenberger thanks again for your contribution.
I hope we can merge it soon

**Review by wing328 — APPROVED** [review]
LGTM. Thanks for the PR.

**Pull request merged**
